### PR TITLE
Refactor annotated HTTP service

### LINKED ISSUE #1175: Refactor annotated HTTP service
The first goal is to extract a common code from `AnnotatedHttpServiceMethod` and `BeanRequestConverterFunction` to support `@Param`, `@Header` and `@Default` annotations in a same way.

It might be to support `@RequestObject` annotation for a bean so that a bean is able to hold another bean. I'm not sure yet if this feature is good or not.

### PR DESCRIPTION
**Pull request by hyangtack**
Motivation:
The first goal is to extract a common code from AnnotatedHttpServiceMethod and BeanRequestConverterFunction to support `@Param`, `@Header` and `@Default` annotations in a same way.
Next, converting a bean in a recursive way would be useful. e.g. a bean inside another bean.

Modifications:
- Add `AnnotatedValueResolver` in order to resolve values in a same way between annotated service method and bean request converter.
- Change `AnnotatedHttpServices` class to `AnnotatedHttpServiceFactory` and refactor it.
- Add `AnnotatedBeanFactory` which replaces `BeanRequestConverterFunction`.
- Now `@Default` and `@RequestObject` can be also put on a field, method and constructor.
- Now `List<?>`, `Set<?>` and `Optional<List<?>>` can be specified with `@Param` and `@Header`.
  - `@Param` is only allowed to HTTP parameters. A path variable cannot be specified as a collection.
  - The element type of a collection is not limited to `String` anymore. The elements would be automatically converted to the desired type.
    e.g. `List<Integer>`, `Set<UserEnumType>`, etc.

Result:
- Closes #1175

### CODE REVIEW

**Comment by hyangtack** [comment]
Let me add negative test cases more.

File: core/src/main/java/com/linecorp/armeria/server/AbstractVirtualHostBuilder.java [hunk 1]
```diff
@@ -446,9 +447,9 @@ public B annotatedService(String pathPrefix, Object service,
         requireNonNull(decorator, "decorator");
         requireNonNull(exceptionHandlersAndConverters, "exceptionHandlersAndConverters");
 
-        final List<AnnotatedHttpService> entries =
-                AnnotatedHttpServices.build(pathPrefix, service, exceptionHandlersAndConverters);
-        entries.forEach(e -> service(e.pathMapping(), decorator.apply(e.decorator().apply(e))));
+        final List<AnnotatedHttpServiceElement> elements =
+                AnnotatedHttpServiceFactory.find(pathPrefix, service, exceptionHandlersAndConverters);
+        elements.forEach(el -> service(el.pathMapping(), decorator.apply(el.decorator().apply(el.service()))));
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AbstractVirtualHostBuilder.java]: `e` or `elem`?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedBeanFactory.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        }
+        return ret;
+    }
+
+    private AnnotatedBeanFactory() {}
+
+    /**
+     * An identifier of the registered bean factory.
+     */
+    static final class BeanId {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedBeanFactory.java]: `BeanFactoryId`?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedBeanFactory.java [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        if (!factories.containsKey(beanId)) {
+            factories.put(beanId, firstNonNull(createFactory(beanId, objectResolvers), unsupportedBeanFactory));
+        }
+        return beanId;
+    }
+
+    /**
+     * Returns a factory of the specified {@link BeanId}.
+     */
+    @Nullable
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedBeanFactory.java]: Not a nullable method? It returns an `Optional`.

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedBeanFactory.java [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+            // A default constructor exists but there is no annotated field or method.
+            return null;
+        }
+
+        // Suppress Java language access checking.
+        constructor.getKey().setAccessible(true);
+        fields.forEach(field -> field.getKey().setAccessible(true));
+        methods.forEach(method -> method.getKey().setAccessible(true));
+
+        logger.info("Register bean factory: {}", beanId);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedBeanFactory.java]: Probably better using `debug()`?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpService.java [hunk 5]
```diff
@@ -16,85 +16,221 @@
...
+        this.exceptionHandlers = ImmutableList.copyOf(
+                requireNonNull(exceptionHandlers, "exceptionHandlers"));
+        this.responseConverters = ImmutableList.copyOf(
+                requireNonNull(responseConverters, "responseConverters"));
+
+        aggregationStrategy = AggregationStrategy.resolve(resolvers);
+
+        final Class<?> returnType = method.getReturnType();
+        returnHttpResponseType = HttpResponse.class.isAssignableFrom(returnType);
+        returnCompletionStageType = CompletionStage.class.isAssignableFrom(returnType);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpService.java]: Perhaps we could use an `enum` and use `switch-case` in `serve0()`?

File: core/src/test/java/com/linecorp/armeria/server/AnnotatedValueResolverTest.java [hunk 6]
```diff
@@ -0,0 +1,959 @@
...
+                arg -> existingHttpHeaders.contains(arg.getArguments()[0]) ? headerValues
+                                                                           : ImmutableList.of());
+    }
+
+    static boolean shouldExistHttpHeader(AnnotatedValueResolver element) {
+        return element.isRequired() ||
+               existingHttpHeaders.contains(AsciiString.of(element.name()));
+    }
+
+    static boolean shouldExistHttpParameter(AnnotatedValueResolver element) {
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/AnnotatedValueResolverTest.java]: `shouldHttpParameterExist`?

File: site/src/sphinx/server-annotated-service.rst [hunk 7]
```diff
@@ -263,7 +263,9 @@ Getting an HTTP header
 
 Armeria also provides :api:`@Header` annotation to inject an HTTP header value into a parameter.
 The parameter annotated with :api:`@Header` can also be specified as one of the built-in types as follows.
-:api:`@Default` and ``Optional<?>`` are also supported.
+:api:`@Default` and ``Optional<?>`` are also supported. Unlike :api:`@Param` annotation, :api:`@Header`
+annotation supports ``List`` or ``Set`` as its type because HTTP headers can be added several times with the
```
Comment (trustin) [inline site/src/sphinx/server-annotated-service.rst]: I'm curious if we need to support `List` or `Set` for query strings, e.g. `?foo=1&foo=2` could be translated into `@Param("foo") List<String>`.

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java [hunk 8]
```diff
@@ -0,0 +1,764 @@
...
+            final String name = findName(param, typeElement);
+            if (pathParams.contains(name)) {
+                final AnnotatedValueResolver element =
+                        ofHttpComponent(param, name, typeElement, type,
+                                        annotatedElement.getAnnotation(Default.class),
+                                        httpPathOrParamValueResolver(ctx -> ctx.context().pathParam(name)),
+                                        false, AggregationStrategy.NONE);
+                if (!element.isRequired()) {
+                    throw new IllegalArgumentException(
+                            "Path variable '" + name + "' should not be an optional.");
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java]: - `Path variable 'foo' should not be an optional.` -> `Path variable 'foo' annotated with @Default` cannot be an Optional.`
- Using `@Default` on an `Optional` parameter is not optimal, but it's probably OK to support such a use case? We may want to log some warning though.

**Comment by hyangtack** [comment]
Will ping @line/cdtf if it's ready for review

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java [hunk 9]
```diff
@@ -0,0 +1,1001 @@
...
+                          .add("resolver", resolver)
+                          .add("enumConverter", enumConverter)
+                          .toString();
+    }
+
+    static Builder builder(AnnotatedElement annotatedElement, Type type) {
+        return new Builder(annotatedElement, type);
+    }
+
+    static class Builder {
```
Comment (hyangtack) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java]: I added a builder for `AnnotatedValueResolver` to make each annotation's specification clear like https://github.com/line/armeria/pull/1191/files#diff-9df84bbcae0217f0942af4e4d6a3f8c0R244.

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java [hunk 10]
```diff
@@ -0,0 +1,1001 @@
...
+            return request;
+        }
+
+        @Nullable
+        AggregatedHttpMessage message() {
+            return message;
+        }
+
+        HttpParameters httpParameters() {
+            // Use a temporary variable to reduce the number of reads of the volatile field.
```
Comment (hyangtack) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java]: Borrowed from https://github.com/apache/commons-lang/blob/master/src/main/java/org/apache/commons/lang3/concurrent/LazyInitializer.java#L97

File: core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceRequestConverterTest.java [hunk 11]
```diff
@@ -755,54 +753,6 @@ public void testDefaultRequestConverter_bean3() throws Exception {
         assertThat(response.content().toStringUtf8()).isEqualTo(expectedResponseContent);
     }
 
-    @Test
-    public void testDefaultRequestConverter_beanDefineError() throws Exception {
```
Comment (hyangtack) [inline core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceRequestConverterTest.java]: Need to add these cases tomorrow

File: site/src/sphinx/server-annotated-service.rst [hunk 12]
```diff
@@ -263,7 +280,9 @@ Getting an HTTP header
 
 Armeria also provides :api:`@Header` annotation to inject an HTTP header value into a parameter.
 The parameter annotated with :api:`@Header` can also be specified as one of the built-in types as follows.
-:api:`@Default` and ``Optional<?>`` are also supported.
+:api:`@Default` and ``Optional<?>`` are also supported. :api:`@Header` annotation also supports
+``List<String>`` or ``Set<String>`` as its type because HTTP headers can be added several times
```
Comment (hyangtack) [inline site/src/sphinx/server-annotated-service.rst]: Need to update tomorrow.

File: core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceRequestConverterTest.java (hunk 11 above)
Comment (hyangtack) [reply core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceRequestConverterTest.java]: Moved to https://github.com/line/armeria/pull/1191/files#diff-acfda6b218745abeaece7a779cd740b7R35

**Comment by hyangtack** [comment]
@line/cdtf Review please. :-)

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java [hunk 13]
```diff
@@ -1,712 +0,0 @@
...
-        this.responseConverters = ImmutableList.copyOf(
-                requireNonNull(responseConverters, "responseConverters"));
-
-        parameters = parameters(method, pathMapping.paramNames(), !requestConverters.isEmpty());
-        final Class<?> returnType = method.getReturnType();
-        isAsynchronous = HttpResponse.class.isAssignableFrom(returnType) ||
-                         CompletionStage.class.isAssignableFrom(returnType);
-        aggregationStrategy = AggregationStrategy.resolve(parameters);
-
-        this.method.setAccessible(true);
```
Comment (kojilin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: I found the problem of CI, so we need to call this in new code.

Comment (kojilin) [reply core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: Maybe need to add test for `IllegalAccess` case.

Comment (hyangtack) [reply core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: Thanks, @kojilin ! I missed that code. 😱

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedBeanFactory.java [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+            // A default constructor exists but there is no annotated field or method.
+            return null;
+        }
+
+        // Suppress Java language access checking.
+        constructor.getKey().setAccessible(true);
+        fields.forEach(field -> field.getKey().setAccessible(true));
+        methods.forEach(method -> method.getKey().setAccessible(true));
+
+        logger.debug("Register bean factory: {}", beanFactoryId);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedBeanFactory.java]: `Registered a bean factory` ?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedBeanFactory.java [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+        };
+    }
+
+    @Nullable
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private static <T> Entry<Constructor<T>, List<AnnotatedValueResolver>> findConstructor(
+            BeanFactoryId beanFactoryId, List<RequestObjectResolver> objectResolvers) {
+
+        Constructor<T> defaultConstructor = null;
+        final List<Entry<Constructor<T>, List<AnnotatedValueResolver>>> candidates = new ArrayList<>();
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedBeanFactory.java]: I think we don't really use a `List` here but just a single value holder:

```java
for (constructor : constructors) {
    ...
    if (!resolvers.isEmpty()) {
        if (candidate == null) {
            candidate = new SimpleImmutableEntry<>(...);
        } else {
            throw new IllegalARgumentException("too many annotated constructors in ...");
        }
    }
}

if (candidate == null) {
    return defaultConstructor != null ? ...;
}

return candidate;
```

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceFactory.java [hunk 16]
```diff
@@ -136,14 +126,139 @@
...
+                                                      List<ResponseConverterFunction> baseResponseConverters) {
+
+        final Set<Annotation> methodAnnotations = httpMethodAnnotations(method);
+        if (methodAnnotations.isEmpty()) {
+            throw new IllegalArgumentException("HTTP Method specification is missing: " + method.getName());
+        }
+
+        final Set<HttpMethod> methods = toHttpMethods(methodAnnotations);
+        if (methods.isEmpty()) {
+            throw new IllegalArgumentException("HTTP Method specification is missing: " + method.getName());
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceFactory.java]: How about:

    method.getDeclaringClass().getName() + '#' + method.getName() +
    " must have an HTTP method annotation."

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceFactory.java [hunk 17]
```diff
@@ -136,14 +126,139 @@
...
+        final Set<String> expectedParamNames = pathMapping.paramNames();
+        final Set<String> requiredParamNames =
+                resolvers.stream()
+                         .filter(AnnotatedValueResolver::isPathVariable)
+                         .map(AnnotatedValueResolver::httpElementName)
+                         .collect(Collectors.toSet());
+
+        if (!expectedParamNames.containsAll(requiredParamNames)) {
+            final Set<String> missing = Sets.difference(requiredParamNames, expectedParamNames);
+            throw new IllegalArgumentException("Missing @Param exists: " + missing);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceFactory.java]: How about:

    "cannot find path variables: " + missing

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java [hunk 18]
```diff
@@ -0,0 +1,1080 @@
...
+            if (value != null) {
+                return value;
+            }
+            throw new IllegalArgumentException("No suitable request converter found for a @" +
+                                               RequestObject.class.getSimpleName() + " '" +
+                                               resolver.elementType().getSimpleName() + '\'');
+        };
+    }
+
+    // Test-purpose.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java]: How about adding `@VisibleForTesting` to `annotataion()` as well?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java (hunk 10 above)
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java]: I think it's too trivial and well known pattern (double-checked locking) and thus there's no need for an attribution.

File: core/src/test/java/com/linecorp/armeria/server/AnnotatedValueResolverTest.java [hunk 19]
```diff
@@ -0,0 +1,1023 @@
...
+    @BeforeClass
+    public static void setUp() throws Exception {
+        when(resolverContext.httpParameters()).thenReturn(parameters);
+        when(resolverContext.context()).thenReturn(context);
+        when(resolverContext.request()).thenReturn(request);
+        when(resolverContext.message()).thenReturn(message);
+
+        when(context.request()).thenReturn(request);
+        when(context.query())
+                .thenReturn(new QueryStringEncoder(
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/AnnotatedValueResolverTest.java]: Seems like `new QueryStringEncoder()` is not necessary at all or you need to use `QueryStringEncoder.addParam()` for each parameters?

File: site/src/sphinx/server-annotated-service.rst [hunk 20]
```diff
@@ -242,6 +242,23 @@ as the value of parameter ``name``. If there is no parameter named ``name`` in t
         }
     }
 
+If multiple parameters exist with the same name in a query string, they can be injected as a ``List<?>``
+or ``Set<?>`` type, e.g. ``/hello1?number=1&number=2&number=3``. You can use :api:`@Default` annotation
```
Comment (trustin) [inline site/src/sphinx/server-annotated-service.rst]: How about just `a List<?> or a Set<?>`? (no 'type')

File: site/src/sphinx/server-annotated-service.rst [hunk 21]
```diff
@@ -263,7 +280,8 @@ Getting an HTTP header
 
 Armeria also provides :api:`@Header` annotation to inject an HTTP header value into a parameter.
 The parameter annotated with :api:`@Header` can also be specified as one of the built-in types as follows.
-:api:`@Default` and ``Optional<?>`` are also supported.
+:api:`@Default` and ``Optional<?>`` are also supported. :api:`@Header` annotation also supports
+``List<?>`` or ``Set<?>`` as its type because HTTP headers can be added several times with the same name.
```
Comment (trustin) [inline site/src/sphinx/server-annotated-service.rst]: Ditto

File: site/src/sphinx/server-annotated-service.rst [hunk 22]
```diff
@@ -455,7 +479,7 @@ converting ``MyRequest``.
 Injecting value of parameters and HTTP headers into a Java object
 """""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
 
-:api:`BeanRequestConverterFunction` is a built-in request converter for Java object. You can use it by
+There is a built-in request converter for Java object. You can use it by
```
Comment (trustin) [inline site/src/sphinx/server-annotated-service.rst]: - There is a built-in -> Armeria provides a generic built-in
- Java object. You can use it by -> Java objects which is activated by

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java [hunk 23]
```diff
@@ -0,0 +1,1079 @@
...
+    }
+
+    boolean isAnnotationType(Class<? extends Annotation> target) {
+        return annotation != null &&
+               annotation.annotationType() == target;
+    }
+
+    @Nullable
+    String httpElementName() {
+        // Currently, this is non-null only if the element one of the HTTP path variable, parameter or header.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java]: element is?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java [hunk 24]
```diff
@@ -0,0 +1,1079 @@
...
+                                                 : null;
+    }
+
+    @Nullable
+    private static EnumConverter<?> enumConverter(Class<?> elementType) {
+        if (!elementType.isEnum()) {
+            return null;
+        }
+        return enumConverters.computeIfAbsent(elementType, newElementType -> {
+            logger.info("Register Enum {}", newElementType);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java]: Registered an Enum?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java [hunk 25]
```diff
@@ -0,0 +1,1079 @@
...
+        /**
+         * Sets whether the element can be annotated with {@link Default} annotation.
+         */
+        private Builder supportDefault(boolean supportDefault) {
+            this.supportDefault = supportDefault;
+            return this;
+        }
+
+        /**
+         * Sets an {@link AnnotatedElement} which is to infer its type.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java]: is used?

File: site/src/sphinx/server-annotated-service.rst [hunk 26]
```diff
@@ -242,6 +242,23 @@ as the value of parameter ``name``. If there is no parameter named ``name`` in t
...
+or ``Optional<?>`` class here, too.
+
+.. code-block:: java
+
+    public class MyAnnotatedService {
+        @Get("/hello1")
+        public HttpResponse hello1(@Param("number") List<Integer> numbers) { ... }
+
+        @Get("/hello2")
+        public HttpResponse hello2(@Param("number") @Default("1") List<Integer> numbers) { ... }
```
Comment (minwoox) [inline site/src/sphinx/server-annotated-service.rst]: Could you add a comment how the default value gets into the `numbers`?

**Pull request merged**